core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala

```diff
@@ -295,7 +298,20 @@ class DAGSchedulerSuite extends SparkFunSuite with LocalSparkContext with TimeLi
 
   override def beforeEach(): Unit = {
     super.beforeEach()
-    init(new SparkConf())
+  }
+
+  override protected def test(testName: String, testTags: Tag*)(testFun: => Any)
+      (implicit pos: Position): Unit = {
+    testWithSparkConf(testName, testTags: _*)()(testFun)(pos)
+  }
+
+  private def testWithSparkConf(testName: String, testTags: Tag*)
```

Comment:
If I understand correctly, you are doing the hack here because you need to modify the SparkConf within `beforeEach` (between `super.beforeEach()` and `init()`). In other words, you don't need to call `testWithSparkConf` instead of `test`, if you don't do extra initialization at the `beforeEach()` stage. Thus, this change is necessarily useful for those test suites you listed, right?

A more staightforward idea would likely to be having a `withConfig(pairs: (String, String)*)` method to create a new SparkConf with the specified configuration values？ The idea doesn't simplify `DAGSchedulerSuite` that much, because you still need to first stop the SparkContext then call `init()` with your own SparkConf, but at least it's not worse than the current approach, and more easy to understand and to be reused.

Reply:
For your first question, yes, it is.
For your second question, `withConfig(pairs: (String, String)*)` mean to stop the `SparkContext`  and then call `init()`. So I created the function `testWithSparkConf` which avoid to stop the `SparkContext` then call `init()`.

Reply:
I tend to agree with @jiangxb1987 .  There're only 7 places using the `testWithSparkConf()` comparing to other 81 tests within the `DAGSchedulerSuite`. We could just call `init` inside each test to ease other tests who needs to call `afterEach` inside the `test`.

And I have another idea for the whole thing. That is, we could probably initialize the `SparkContext` like this:


```scala
trait LocalSparkContext ... {
  @transient private var _sc: SparkContext = _
   private val conf = new SparkConf()

  def sc: SparkContext = {
      if (_sc == null) {
        _sc = new SparkContext(conf)
      }
      _sc
  }

  def withConf(pairs: (String, String)*) = {
    if (_sc != null) { 
       // probably, log warning when SparkContext already initialized
       // since configurations won't take effect in this case
     }
     paris.foreach { case (k, v) => conf.set(k, v) }
  }

  override def afterEach(): Unit = {
    try {
      resetSparkContext()
    } finally {
      super.afterEach()
    }
  }

  def resetSparkContext(): Unit = {
    LocalSparkContext.stop(sc)
    ResourceProfile.clearDefaultProfile()
    sc = null
  }
}
```

@beliefer @jiangxb1987 WDYT?

Reply:
I tend to agree with @jiangxb1987 .  There're only 7 places using the `testWithSparkConf()` comparing to other 81 tests within the `DAGSchedulerSuite`. We could just call `init` inside each test to ease other tests who needs to call `afterEach` inside the `test`.

And I have another idea for the whole thing. That is, we could probably initialize the `SparkContext` like this:


```scala
trait LocalSparkContext ... {
  @transient private var _sc: SparkContext = _
   private val conf = new SparkConf()

  def sc: SparkContext = {
      if (_sc == null) {
        _sc = new SparkContext(conf)
      }
      _sc
  }

  def withConf(pairs: (String, String)*) = {
    if (_sc != null) { 
       // probably, log warning when SparkContext already initialized
       // since configurations won't take effect in this case
     }
     paris.foreach { case (k, v) => conf.set(k, v) }
  }

  override def afterEach(): Unit = {
    try {
      resetSparkContext()
    } finally {
      super.afterEach()
    }
  }

  def resetSparkContext(): Unit = {
    LocalSparkContext.stop(sc)
    ResourceProfile.clearDefaultProfile()
    sc = null
  }
}
```

@beliefer @jiangxb1987 WDYT?

Reply:
+1 to the above proposal

Reply:
OK. I will implement this proposal.